### PR TITLE
Function metrics

### PR DESCRIPTION
**Pull request by vlaksi**
I add C# documentation in this form. For someone who is new in this project to know what is some metric and how someone else call her.

image

### CODE REVIEW

**Comment by vlaksi** [comment]
- Added unit tests for NOP & NOLV with manual calculate for metrics

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpMetricCalculator.cs [hunk 1]
```diff
@@ -77,10 +79,24 @@ public CaDETMemberMetrics CalculateMemberMetrics(MemberDeclarationSyntax member)
...
+        private int GetNumberOfLocalVariables(MemberDeclarationSyntax method)
+        {
+            // TODO: Change this also to real value
+            return 10;
+        }
+
+        private int GetNumberOfParameters(MemberDeclarationSyntax method)
+        {
+            // First because of structure type of 'OfType<ParameterListSyntax>()'
+            return method.DescendantNodes().OfType<ParameterListSyntax>().First().Parameters.Count;
```
Comment (vlaksi) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpMetricCalculator.cs:97]: - I find this way to get number of parameters
- There is some better i guess, so fell free to inform me about it

**Comment by vlaksi** [comment]
## Issue with unit testing

  - I tested ONLY my functions and they work
  - After testing my functions, i just run all other tests and there is issue

image


image

## Problem

Problem is when i include NOP & NOLV in CaDETMemberMetrics class.

image

Without them all other tests works(except my tests).

image


image




@Luburic & @Pufke  have you some fix for this ?

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpMetricCalculator.cs (hunk 1 above)
Comment (Luburic) [reply RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpMetricCalculator.cs:94]: This seems alright, but we should make a good suite of test data to see if the results of the calculation are as expected.

The main issue (the reason why other tests are failing) is that you do not handle the edge case with properties that lack parameter lists (even empty ones). Since we are calculating metrics for all types of executable methods (properties, constructors, methods) we need to check them as well.

Try replacing the body with this code:
            var paramLists = method.DescendantNodes().OfType<ParameterListSyntax>().ToList();
            return !paramLists.Any() ? 0 : paramLists.First().Parameters.Count;

Two things to consider:
- What is the situation when we have multiple parameter lists?
- Pro tip: Run all unit tests before committing anything :)

File: RepositoryCompilerTests/Unit/CaDETCodeModelTests.cs [hunk 2]
```diff
@@ -125,5 +126,34 @@ public void Determines_if_is_data_class()
...
+
+        [Fact]
+        public void Calculates_number_of_parameters()
+        {
+            CodeModelBuilder builder = new CodeModelBuilder(LanguageEnum.CSharp);
+
+            List<CaDETClass> classes = builder.BuildCodeModel(_testDataFactory.GetGitAdapterClassText());
+
+            var gitClass = classes.First();
+            gitClass.Methods.Find(method => method.Name.Equals("CheckForNewCommits")).Metrics.NOP.ShouldBe(0);
```
Comment (Luburic) [inline RepositoryCompilerTests/Unit/CaDETCodeModelTests.cs:138]: We should create a helper function for finding methods that makes the code more DRY, e.g.

FindMethod(CaDETClass c, String methodName)
{
  return c.Methods.Find(method => method.Name.Equals(methodName));
}

We can then resolve Feature Envy and reduce the parameter count by moving the method to the CaDETClass.

**Review by Luburic — CHANGES_REQUESTED** [review]
You will find the comments inline, including the bugfix. We can discuss the review on Thursday.

**Comment by Luburic** [comment]
[Function metrics](https://trello.com/c/nLtmUBJK/72-function-metrics)

**Review by Luburic — COMMENTED** [review]
Fixes up to this point look good. When you move the parameter calculation to the Member class and refactor the # of params metric calculation make sure to run all unit tests before the commit.

**Review by Luburic — APPROVED** [review]
Lgtm.
